File: LeetCode 0348 Design Tic-Tac-Toe.py
class TicTacToe(object): # 85% Time, 66% RAM

    def __init__(self, n):
        """
        Initialize your data structure here.
        :type n: int
        """
        self.rows = [0] * n # 所有行分别的元素和
        self.cols = [0] * n # 所有列分别的元素和
        self.diag = 0 # 对角线元素和
        self.adia = 0 # 反对角线元素和
        self.n = n

    def move(self, row, col, player):
        """
        Player {player} makes a move at ({row}, {col}).
        @param row The row of the board.
        @param col The column of the board.
        @param player The player, can be either 1 or 2.
        @return The current winning condition, can be either:
                0: No one wins.
                1: Player 1 wins.
                2: Player 2 wins.
        :type row: int
        :type col: int
        :type player: int
        :rtype: int
        """
        # 插入值
        p = player * 2 - 3 # offset, p1 -> -1, p2 -> 1
        self.rows[row] += p # 更新行和
        self.cols[col] += p # 更新列和
        if col == row: self.diag += p # 更新对角线和
        if col + row + 1 == self.n: self.adia += p # 更新反对角线和
        
        if self.n in [self.rows[row], self.cols[col], self.diag, self.adia]: return 2 # 任意和达到n，则该项全是“1”，返回2
        if -self.n in [self.rows[row], self.cols[col], self.diag, self.adia]: return 1 # 任意和达到n则全是“-1”，返回1
        return 0
        
# Your TicTacToe object will be instantiated and called as such:
# obj = TicTacToe(n)
# param_1 = obj.move(row,col,player)

File: LeetCode 0348 Design Tic-tac-toe.java
class TicTacToe { // 99, 9
    int n;
    int[] rows;
    int[] cols;
    int diag;
    int adiag;

    /** Initialize your data structure here. */
    public TicTacToe(int n) {
        this.n = n;
        this.rows = new int[n];
        this.cols = new int[n];
        this.diag = 0;
        this.adiag = 0;
        Arrays.fill(rows, 0);
        Arrays.fill(cols, 0);
    }
    
    /** Player {player} makes a move at ({row}, {col}).
        @param row The row of the board.
        @param col The column of the board.
        @param player The player, can be either 1 or 2.
        @return The current winning condition, can be either:
                0: No one wins.
                1: Player 1 wins.
                2: Player 2 wins. */
    public int move(int row, int col, int player) {
        int step = player == 2 ? 1 : -1;
        this.rows[row] += step;
        this.cols[col] += step;
        if (row == col)
            this.diag += step;
        if (row + col == n - 1)
            this.adiag += step;
        
        if (this.rows[row] == n || this.cols[col] == n || this.diag == n || this.adiag == n)
            return 2;
        if (this.rows[row] == -n || this.cols[col] == -n || this.diag == -n || this.adiag == -n)
            return 1;
        return 0;
    }
}

/**
 * Your TicTacToe object will be instantiated and called as such:
 * TicTacToe obj = new TicTacToe(n);
 * int param_1 = obj.move(row,col,player);
 */